libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -177,39 +178,73 @@ namespace rttr { namespace mapGenerator {
     {
         const auto& textures = map.textureMap;
         auto& resources = map.resources;
+        struct MineableResourceInfo
+        {
+            // The current "budget", that is, how many many are we over, or under, compared to the desired distribution.
+            int budget = 0;
+            // The ratio is the user defined ratio for this resource.
+            int ratio;
+            // The resource encoding
+            libsiedler2::Resource resource;
+            MineableResourceInfo(int ratio, libsiedler2::Resource resource) : ratio(ratio), resource(resource) {}
+        };
+        MineableResourceInfo mRIs[4] = {MineableResourceInfo(settings.ratioCoal, libsiedler2::R_Coal),
+                                        MineableResourceInfo(settings.ratioGold, libsiedler2::R_Gold),
+                                        MineableResourceInfo(settings.ratioIron, libsiedler2::R_Iron),
+                                        MineableResourceInfo(settings.ratioGranite, libsiedler2::R_Granite)};
         int total = settings.ratioCoal + settings.ratioGold + settings.ratioIron + settings.ratioGranite;
 
+        // Helper to pick an index [0,4) into the mRIs array that identifies a randomly selected resource we should
+        // place. Note that we do not perform budget adjustments and checks here.
+        auto resourcePicker = [&]() {
+            int ratio = 0;
+            int randomNumber = rnd.RandomValue(1, std::max(1, total));
+            for(int i = 0; i < 4; ++i)
+            {
+                ratio += mRIs[i].ratio;
+                if(randomNumber < ratio)
```

Comment:
I think this should be
```suggestion
                if(randomNumber <= ratio)
```

Example: All ratios are 20, total is 100 --> 1-20 is first, 21-40 is 2nd, ... 81 to 100 is 4th. So 20 values per index
Currently if randomNumber is 100 no value will be picked and (at least) the first would only have 19 "chances" while the 2nd has 20